File: status_monitor.py
import json
import threading
from pathlib import Path
from typing import Callable, Optional

from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler, FileSystemEvent

from config import CopilotStatus, COPILOT_STATE_DIR, POLL_INTERVAL

_STATUS_MAP = {
    "assistant.turn_start": CopilotStatus.BUSY,
    "tool.execution_start": CopilotStatus.BUSY,
    "assistant.message": CopilotStatus.BUSY,
    "hook.start": CopilotStatus.BUSY,
    "assistant.turn_end": CopilotStatus.WAITING,
    "session.task_complete": CopilotStatus.IDLE,
}


class _EventHandler(FileSystemEventHandler):
    """Watches for changes to .jsonl and .lock files."""

    def __init__(self, callback: Callable[[], None]):
        super().__init__()
        self._callback = callback

    def on_modified(self, event: FileSystemEvent) -> None:
        if not event.is_directory and self._is_relevant(event.src_path):
            self._callback()

    def on_created(self, event: FileSystemEvent) -> None:
        if not event.is_directory and self._is_relevant(event.src_path):
            self._callback()

    @staticmethod
    def _is_relevant(path: str) -> bool:
        return path.endswith(".jsonl") or path.endswith(".lock")


class StatusMonitor:
    """Monitors Copilot CLI session state to determine current status."""

    def __init__(self, on_status_change: Callable[[CopilotStatus, str], None]):
        self._on_status_change = on_status_change
        self._last_status: CopilotStatus = CopilotStatus.IDLE
        self._last_text: str = ""
        self._last_model_name: str = ""
        self._last_file_size: int = 0
        self._lock = threading.Lock()
        self._stop_event = threading.Event()
        self._observer: Optional[Observer] = None
        self._poll_thread: Optional[threading.Thread] = None
        self._debounce_timer: Optional[threading.Timer] = None

    # -- public API --

    def start(self) -> None:
        """Begin monitoring with watchdog + polling fallback."""
        self._stop_event.clear()

        # Initial status check
        self._check_and_notify()

        # Watchdog observer
        self._start_observer()

        # Polling fallback
        self._poll_thread = threading.Thread(
            target=self._poll_loop, daemon=True, name="status-poll"
        )
        self._poll_thread.start()

    def stop(self) -> None:
        """Stop monitoring and clean up."""
        self._stop_event.set()

        if self._debounce_timer is not None:
            self._debounce_timer.cancel()

        if self._observer is not None:
            try:
                self._observer.stop()
                self._observer.join(timeout=2)
            except Exception:
                pass
            self._observer = None

        if self._poll_thread is not None:
            self._poll_thread.join(timeout=3)
            self._poll_thread = None

    def get_current_status(self) -> tuple[CopilotStatus, str]:
        """Synchronously check and return current status and text."""
        with self._lock:
            session = self._find_active_session()
            if session is None:
                return (CopilotStatus.IDLE, "")
            return self._parse_status(session)

    @property
    def model_name(self) -> str:
        """Last detected model name (e.g. 'claude-opus-4.6')."""
        return self._last_model_name

    @property
    def context_bytes(self) -> int:
        """Size of current session's events.jsonl in bytes."""
        return self._last_file_size

    # -- internal --

    def _start_observer(self) -> None:
        if not COPILOT_STATE_DIR.exists():
            print(
                f"[StatusMonitor] State dir not found: {COPILOT_STATE_DIR} "
                "— relying on polling"
            )
            return
        try:
            self._observer = Observer()
            handler = _EventHandler(self._debounced_check)
            self._observer.schedule(handler, str(COPILOT_STATE_DIR), recursive=True)
            self._observer.daemon = True
            self._observer.start()
        except Exception as exc:
            print(f"[StatusMonitor] Failed to start watchdog: {exc}")
            self._observer = None

    def _debounced_check(self) -> None:
        """Schedule a status check after a short debounce window."""
        if self._debounce_timer is not None:
            self._debounce_timer.cancel()
        self._debounce_timer = threading.Timer(0.15, self._check_and_notify)
        self._debounce_timer.daemon = True
        self._debounce_timer.start()

    def _poll_loop(self) -> None:
        interval = POLL_INTERVAL / 1000.0
        while not self._stop_event.is_set():
            self._stop_event.wait(interval)
            if not self._stop_event.is_set():
                self._check_and_notify()

    def _check_and_notify(self) -> None:
        with self._lock:
            session = self._find_active_session()
            if session:
                new_status, new_text = self._parse_status(session)
            else:
                new_status = CopilotStatus.IDLE
                new_text = ""
                self._last_file_size = 0
            if new_status != self._last_status or new_text != self._last_text:
                self._last_status = new_status
                self._last_text = new_text
                try:
                    self._on_status_change(new_status, new_text)
                except Exception as exc:
                    print(f"[StatusMonitor] Callback error: {exc}")

    def _find_active_session(self) -> Optional[Path]:
        """Return the session dir with a lock file and the newest events.jsonl."""
        try:
            if not COPILOT_STATE_DIR.exists():
                return None
        except (OSError, PermissionError):
            return None

        best: Optional[Path] = None
        best_mtime: float = 0.0

        try:
            entries = list(COPILOT_STATE_DIR.iterdir())
        except (OSError, PermissionError):
            return None

        for entry in entries:
            try:
                if not entry.is_dir():
                    continue
                # Check for at least one inuse.*.lock file
                if not any(entry.glob("inuse.*.lock")):
                    continue
                events_file = entry / "events.jsonl"
                if not events_file.exists():
                    continue
                mtime = events_file.stat().st_mtime
                if mtime > best_mtime:
                    best_mtime = mtime
                    best = entry
            except (OSError, PermissionError):
                continue

        return best

    def _parse_status(self, session_dir: Path) -> tuple[CopilotStatus, str]:
        """Parse the tail of events.jsonl to determine status, intent text,
        model name, and file size (context usage proxy)."""
        events_file = session_dir / "events.jsonl"
        try:
            size = events_file.stat().st_size
            self._last_file_size = size
            if size == 0:
                return (CopilotStatus.IDLE, "")

            with open(events_file, "rb") as fh:
                seek_pos = max(0, size - 8192)
                fh.seek(seek_pos)
                tail = fh.read().decode("utf-8", errors="replace")

            lines = tail.splitlines()

            # If we seeked into the middle, the first line is likely partial
            if seek_pos > 0 and lines:
                lines = lines[1:]

            status = CopilotStatus.IDLE
            status_text = ""
            found_status = False

            for line in reversed(lines):
                line = line.strip()
                if not line:
                    continue
                try:
                    event = json.loads(line)
                except (json.JSONDecodeError, ValueError):
                    continue
                event_type = event.get("type", "")
                data = event.get("data", {})

                # Find status (first match walking backwards)
                if not found_status and event_type in _STATUS_MAP:
                    status = _STATUS_MAP[event_type]
                    found_status = True

                # Find latest intent text (may be before the status event)
                if not status_text:
                    if event_type == "tool.execution_start" and data.get("toolName") == "report_intent":
                        status_text = data.get("arguments", {}).get("intent", "")

                # Pick up model name from session.start or tool.execution_complete
                if not self._last_model_name:
                    if event_type == "session.start":
                        self._last_model_name = data.get("selectedModel", "")
                    elif event_type == "tool.execution_complete" and data.get("model"):
                        self._last_model_name = data.get("model", "")

                # Stop once we have everything
                if found_status and status_text and self._last_model_name:
                    break

            # If we still don't have a model name, do a forward scan of just
            # the first few lines (session.start is typically near the top)
            if not self._last_model_name:
                self._scan_for_model(events_file)

        except FileNotFoundError:
            pass
        except (PermissionError, OSError, UnicodeDecodeError) as exc:
            print(f"[StatusMonitor] Error reading {events_file}: {exc}")

        return (status, status_text)

    def _scan_for_model(self, events_file: Path) -> None:
        """Read the head of events.jsonl looking for session.start model info."""
        try:
            with open(events_file, "r", encoding="utf-8", errors="replace") as fh:
                for line in fh:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        event = json.loads(line)
                    except (json.JSONDecodeError, ValueError):
                        continue
                    if event.get("type") == "session.start":
                        model = event.get("data", {}).get("selectedModel", "")
                        if model:
                            self._last_model_name = model
                        return
                    # Only scan first 20 lines
                    if fh.tell() > 4096:
                        return
        except (OSError, PermissionError):
            pass

File: cpp/src/status_monitor.cpp
#include "status_monitor.h"

#include <efsw/efsw.hpp>
#include <nlohmann/json.hpp>

#include <filesystem>
#include <fstream>
#include <sstream>
#include <iostream>
#include <chrono>
#include <thread>
#include <unordered_map>
#include <cstdlib>

namespace fs = std::filesystem;
using json = nlohmann::json;

// ---------------------------------------------------------------------------
// Event-type → status mapping (mirrors Python _STATUS_MAP)
// ---------------------------------------------------------------------------
static const std::unordered_map<std::string, CopilotStatus> STATUS_MAP = {
    {"assistant.turn_start",  CopilotStatus::BUSY},
    {"tool.execution_start",  CopilotStatus::BUSY},
    {"assistant.message",     CopilotStatus::BUSY},
    {"hook.start",            CopilotStatus::BUSY},
    {"assistant.turn_end",    CopilotStatus::WAITING},
    {"session.task_complete", CopilotStatus::IDLE},
};

// ---------------------------------------------------------------------------
// File-watch listener — routes relevant fs events back via a std::function
// callback, with a 150 ms debounce guard.
// ---------------------------------------------------------------------------
class StatusFileListener : public efsw::FileWatchListener {
public:
    explicit StatusFileListener(std::function<void()> callback)
        : m_callback(std::move(callback)), m_last_ms(0) {}

    void handleFileAction(efsw::WatchID /*id*/, const std::string& /*dir*/,
                          const std::string& filename, efsw::Action action,
                          std::string /*old*/) override {
        // Ignore deletes — the poll loop handles session-end detection
        if (action == efsw::Actions::Delete) return;
        if (ends_with(filename, ".jsonl") || ends_with(filename, ".lock"))
            debounce();
    }

private:
    static bool ends_with(const std::string& s, const std::string& suffix) {
        if (s.size() < suffix.size()) return false;
        return s.compare(s.size() - suffix.size(), suffix.size(), suffix) == 0;
    }

    void debounce() {
        int64_t now_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
            std::chrono::steady_clock::now().time_since_epoch()).count();
        int64_t last = m_last_ms.load(std::memory_order_relaxed);
        if (now_ms - last < 150) return;
        m_last_ms.store(now_ms, std::memory_order_relaxed);
        m_callback();
    }

    std::function<void()> m_callback;
    std::atomic<int64_t>  m_last_ms;
};

// ---------------------------------------------------------------------------
// Module-scope listener storage.
// There is exactly one StatusMonitor instance per process; a static pointer
// lets us attach the listener's lifetime to start()/stop() without changing
// the header (which fixes the unique_ptr<FileWatcher> type to the default
// deleter, preventing a custom-deleter bundle).
// ---------------------------------------------------------------------------
static StatusFileListener* s_listener = nullptr;

// ---------------------------------------------------------------------------
// StatusMonitor
// ---------------------------------------------------------------------------

StatusMonitor::StatusMonitor(Callback on_change)
    : m_on_change(std::move(on_change))
{
    const char* home = std::getenv("HOME");
    m_state_dir = std::string(home ? home : "") + "/.copilot/session-state";
}

StatusMonitor::~StatusMonitor() {
    stop();
}

void StatusMonitor::start() {
    if (m_running.exchange(true)) return; // already running

    check_and_notify();

    if (fs::exists(m_state_dir)) {
        if (s_listener) { delete s_listener; s_listener = nullptr; }
        s_listener = new StatusFileListener([this]() { check_and_notify(); });

        m_watcher = std::make_unique<efsw::FileWatcher>();
        m_watcher->addWatch(m_state_dir, s_listener, /*recursive=*/true);
        m_watcher->watch();
    } else {
        std::cerr << "[StatusMonitor] State dir not found: " << m_state_dir << "\n";
    }

    m_poll_thread = std::make_unique<std::thread>(&StatusMonitor::poll_loop, this);
}

void StatusMonitor::stop() {
    if (!m_running.exchange(false)) return; // already stopped

    // Destroy watcher — stops its background thread before we clean up listener
    m_watcher.reset();
    if (s_listener) { delete s_listener; s_listener = nullptr; }

    if (m_poll_thread && m_poll_thread->joinable()) {
        m_poll_thread->join();
        m_poll_thread.reset();
    }
}

// ---------------------------------------------------------------------------
// Thread-safe public getters
// ---------------------------------------------------------------------------

CopilotStatus StatusMonitor::status() const {
    return m_status.load();
}

std::string StatusMonitor::status_text() const {
    std::lock_guard<std::mutex> lock(m_mutex);
    return m_status_text;
}

std::string StatusMonitor::model_name() const {
    std::lock_guard<std::mutex> lock(m_mutex);
    return m_model_name;
}

size_t StatusMonitor::context_bytes() const {
    return m_context_bytes.load();
}

// ---------------------------------------------------------------------------
// Background poll loop — sleeps in 50 ms chunks so stop() is responsive
// ---------------------------------------------------------------------------
void StatusMonitor::poll_loop() {
    constexpr int CHUNK_MS = 50;
    int elapsed_ms = 0;
    while (m_running.load()) {
        std::this_thread::sleep_for(std::chrono::milliseconds(CHUNK_MS));
        elapsed_ms += CHUNK_MS;
        if (!m_running.load()) break;
        if (elapsed_ms >= POLL_INTERVAL_MS) {
            elapsed_ms = 0;
            check_and_notify();
        }
    }
}

// ---------------------------------------------------------------------------
// Core check — parse state, fire callback if status or text changed.
// Lock is released BEFORE invoking the callback to prevent deadlock when the
// callback's thread tries to call public getters.
// ---------------------------------------------------------------------------
void StatusMonitor::check_and_notify() {
    CopilotStatus new_status = CopilotStatus::IDLE;
    std::string   new_text;
    bool          changed = false;

    {
        std::lock_guard<std::mutex> lock(m_mutex);

        CopilotStatus prev_status = m_status.load();
        std::string   prev_text   = m_status_text;

        std::string session = find_active_session();
        if (!session.empty()) {
            parse_status(session);
        } else {
            m_status.store(CopilotStatus::IDLE);
            m_status_text.clear();
            m_context_bytes.store(0);
        }

        new_status = m_status.load();
        new_text   = m_status_text;
        changed    = (new_status != prev_status || new_text != prev_text);
    }

    if (changed && m_on_change) {
        try {
            m_on_change(new_status, new_text);
        } catch (const std::exception& e) {
            std::cerr << "[StatusMonitor] Callback error: " << e.what() << "\n";
        }
    }
}

// ---------------------------------------------------------------------------
// find_active_session — returns path of the subdir that:
//   • is a directory
//   • contains at least one inuse.*.lock file
//   • contains events.jsonl
//   • has the most-recent events.jsonl mtime
// Called while m_mutex is already held by check_and_notify().
// ---------------------------------------------------------------------------
std::string StatusMonitor::find_active_session() const {
    std::string best;
    std::filesystem::file_time_type best_mtime{};

    try {
        if (!fs::exists(m_state_dir)) return {};

        for (const auto& entry : fs::directory_iterator(m_state_dir)) {
            try {
                if (!entry.is_directory()) continue;

                // Look for inuse.*.lock
                bool has_lock = false;
                for (const auto& sub : fs::directory_iterator(entry.path())) {
                    const std::string fname = sub.path().filename().string();
                    // Pattern: "inuse." + at least one char + ".lock"
                    if (fname.size() >= 12 &&
                        fname.rfind("inuse.", 0) == 0 &&
                        fname.size() > 6 + 5 &&
                        fname.substr(fname.size() - 5) == ".lock") {
                        has_lock = true;
                        break;
                    }
                }
                if (!has_lock) continue;

                fs::path events_file = entry.path() / "events.jsonl";
                if (!fs::exists(events_file)) continue;

                auto mtime = fs::last_write_time(events_file);
                if (best.empty() || mtime > best_mtime) {
                    best_mtime = mtime;
                    best = entry.path().string();
                }
            } catch (...) { continue; }
        }
    } catch (...) { return {}; }

    return best;
}

// ---------------------------------------------------------------------------
// parse_status — tail-reads the last TAIL_READ_BYTES of events.jsonl and
// walks lines in reverse to determine status, intent text, and model name.
// Called while m_mutex is already held.
// ---------------------------------------------------------------------------
void StatusMonitor::parse_status(const std::string& session_dir) {
    fs::path events_file = fs::path(session_dir) / "events.jsonl";

    try {
        auto file_size = fs::file_size(events_file);
        m_context_bytes.store(file_size);

        if (file_size == 0) {
            m_status.store(CopilotStatus::IDLE);
            m_status_text.clear();
            return;
        }

        // Read last TAIL_READ_BYTES
        std::ifstream fh(events_file, std::ios::binary);
        if (!fh.is_open()) return;

        size_t seek_pos = (file_size > TAIL_READ_BYTES)
                          ? static_cast<size_t>(file_size - TAIL_READ_BYTES) : 0;
        fh.seekg(static_cast<std::streamoff>(seek_pos));

        std::string tail((std::istreambuf_iterator<char>(fh)),
                          std::istreambuf_iterator<char>());
        fh.close();

        // Split into lines
        std::vector<std::string> lines;
        {
            std::istringstream ss(tail);
            std::string line;
            while (std::getline(ss, line)) {
                if (!line.empty() && line.back() == '\r') line.pop_back();
                lines.push_back(std::move(line));
            }
        }

        // Discard partial leading line when we seeked into the middle
        if (seek_pos > 0 && !lines.empty())
            lines.erase(lines.begin());

        CopilotStatus status      = CopilotStatus::IDLE;
        std::string   status_text;
        bool          found_status = false;
        bool          found_text   = false;

        for (auto it = lines.rbegin(); it != lines.rend(); ++it) {
            const std::string& line = *it;
            if (line.empty()) continue;

            json event;
            try { event = json::parse(line); } catch (...) { continue; }

            const std::string event_type = event.value("type", std::string{});
            const json        data       = event.value("data", json::object());

            if (!found_status) {
                auto sit = STATUS_MAP.find(event_type);
                if (sit != STATUS_MAP.end()) {
                    status = sit->second;
                    found_status = true;
                }
            }

            if (!found_text && event_type == "tool.execution_start") {
                if (data.value("toolName", std::string{}) == "report_intent") {
                    auto args = data.value("arguments", json::object());
                    std::string intent = args.value("intent", std::string{});
                    if (!intent.empty()) {
                        status_text = std::move(intent);
                        found_text  = true;
                    }
                }
            }

            if (m_model_name.empty()) {
                if (event_type == "session.start") {
                    m_model_name = data.value("selectedModel", std::string{});
                } else if (event_type == "tool.execution_complete") {
                    std::string model = data.value("model", std::string{});
                    if (!model.empty()) m_model_name = std::move(model);
                }
            }

            if (found_status && found_text && !m_model_name.empty()) break;
        }

        m_status.store(status);
        m_status_text = std::move(status_text);

        if (m_model_name.empty())
            scan_for_model(events_file.string());

    } catch (const std::exception& e) {
        std::cerr << "[StatusMonitor] parse_status error: " << e.what() << "\n";
    } catch (...) {}
}

// ---------------------------------------------------------------------------
// scan_for_model — reads from the top of events.jsonl (up to 4 KB) looking
// for a session.start event that carries selectedModel.
// Called while m_mutex is already held.
// ---------------------------------------------------------------------------
void StatusMonitor::scan_for_model(const std::string& events_path) {
    try {
        std::ifstream fh(events_path);
        if (!fh.is_open()) return;

        std::string line;
        while (std::getline(fh, line)) {
            if (!line.empty() && line.back() == '\r') line.pop_back();
            if (line.empty()) continue;

            json event;
            try { event = json::parse(line); } catch (...) { continue; }

            if (event.value("type", std::string{}) == "session.start") {
                auto data  = event.value("data", json::object());
                std::string model = data.value("selectedModel", std::string{});
                if (!model.empty()) m_model_name = std::move(model);
                return;
            }

            if (fh.tellg() > 4096) return;
        }
    } catch (...) {}
}
